Introduce NewMultiCfIterator API

jaykorean opened the pull request:
This PR introduces a new implementation of `Iterator` via a new public API called `NewMultiCfIterator()`. The new API takes a vector of column family handles to build a cross-column-family iterator, which internally maintains multiple `DBIter`s as child iterators from a consistent database state. When a key exists in multiple column families, the iterator selects the value (and wide columns) from the first column family containing the key, following the order provided in the `column_families` parameter. Similar to the merging iterator, a min heap is used to iterate across the child iterators. Backward iteration and direction change functionalities will be implemented in future PRs.

The comparator used to compare keys across different column families will be derived from the iterator of the first column family specified in `column_families`. This comparator will be checked against the comparators from all other column families that the iterator will traverse. If there's a mismatch with any of the comparators, the initialization of the iterator will fail.

Please note that this PR is not enough for users to start using `MultiCfIterator`. The `MultiCfIterator` and related APIs are still marked as "**DO NOT USE - UNDER CONSTRUCTION**". This PR is just the first of many PRs that will follow soon.

This PR includes the following:
- Introduction and partial implementation of the `MultiCfIterator`, which implements the generic `Iterator` interface. The implementation includes the construction of the iterator, `SeekToFirst()`, `Next()`, `Valid()`, `key()`, `value()`, and `columns()`.
- Unit tests to verify iteration across multiple column families in two distinct scenarios: (1) keys are unique across all column families, and (2) the same keys exist in multiple column families.

jaykorean commented on db/multi_cf_iterator_impl.cc:
```diff
@@ -0,0 +1,206 @@
+    assert(Valid());
+
+    auto* current = min_heap_.top().iterator;
+    std::string current_key_copy =
```
Here, we first store the key from the iterator at the top of the heap. Then the top iterator moves to the next element and heap gets re-heapified. If the same key appears at the top of the heap, it repeats. Basically, this Next() finds the next key in order across all the iterators in the heap by skipping the same key that we've seen.

I'd love to see if there's a better way to achieve this (e.g. without copying the first key)

jaykorean commented on db/multi_cf_iterator_test.cc:
```diff
@@ -0,0 +1,198 @@
+  }
+}
+
+TEST_F(MultiCfIteratorTest, DISABLED_IterateAttributeGroups) {
```
Will re-enable it when `columns()` and `attribute_groups()` are implemented. Just to illustrate what they would return when implemented.

jaykorean commented on db/multi_cf_iterator_impl.cc:
```diff
@@ -0,0 +1,51 @@
+  assert(Valid());
+
+  auto* current = min_heap_.top().iterator;
+  std::string current_key_copy =
```
Here, we first store the key from the iterator at the top of the heap. Then the top iterator moves to the next element and heap gets re-heapified. If the same key appears at the top of the heap, it repeats. Basically, this Next() finds the next key in order across all the iterators in the heap by skipping the same key that we've seen.

I'd love to see if there's a better way to achieve this (e.g. without copying the first key)

pdillinger replied:
It seems like you should be able to keep one iterator at the current key (remove from heap), use that key Slice to ensure all others have iterated past it (until top of heap is past it), then advance the first iterator past that key and re-place on the heap.

cbi42 commented on include/rocksdb/db.h:
```diff
@@ -969,6 +971,16 @@ class DB {
+  virtual MultiCfIterator* NewMultiCfIterator(
+      const ReadOptions& options,
+      const std::vector<ColumnFamilyHandle*>& column_families,
+      const Comparator* comparator = BytewiseComparator()) = 0;
```
It feels dangerous to have a default comparator here. If we ask user to provide a comparator, we should probably specify in API comments that the input comparator should order the keys in a CF the same order as that CF's comparator. Maybe we can just use a given column families' comparator and not let user to specify one.

jaykorean replied:
There's actually a use case where cross-CF key comparator is different from within-CF key comparator. e.g. ZippyDB's metadata to be stored in a separate CF.  The key format may be different, but can be compared with a custom logic.

@ltamasi please let me know if my understanding is wrong here.

ltamasi replied:
... so in short, yeah, I think we can just error out during iterator creation if the comparators don't match.

cbi42 commented on db/multi_cf_iterator_impl.cc:
```diff
@@ -0,0 +1,70 @@
+
+  auto* current = min_heap_.top().iterator;
+  std::string current_key_copy =
+      std::string(current->key().data(), current->key().size());
```
Call current->Next() once before while loop?

jaykorean replied:
if `current->Next()` is called before the loop, the key will be different from the previous key, `comparator_->Compare(current->key(), current_key_copy)` will always return non-zero and we won't get into while loop.

Essentially this `Next()` does the following (maybe I should add this as comment at the top of the function)
1. Get the current key at the top of the heap iterator and remember it
2. Move to the next item in the top iterator
3. if the iterator is still valid, re-heapify. Otherwise, pop it from the heap (which also re-heapify). This may bring the same key that we've seen in 1 to the top of the heap. (Just stored in a different CF)
4. Compare the new top key with the key in 1. If the same, go back to step 2.
5. If the heap is empty, we are done with the iteration.

cbi42 commented on db/multi_cf_iterator_test.cc:
```diff
@@ -0,0 +1,198 @@
+
+  // Merge columns from all CFs that key exists and value is stored as wide
+  // column
+  std::vector<WideColumns> expected_wide_columns = {
```
What happens when a column name is duplicated between CFs here? 

I originally thought `columns()` would behave like `value()` and returns WideColumns from the first CF that contains the key. And user can call `attribute_groups()` if they want results from all CFs. Is there a reason for returning this `expected_wide_columns`?

jaykorean replied:
That's a great question! I've briefly discussed this with @ltamasi. Please let me know what you think.

One idea that I suggested is providing an option to the user to choose the behavior.
1. Only the first one if key exists in multiple CFs
2. Merge them all (But this also has the similar issue - "same column name in multiple wide columns to merge")

However, I also feel that having more options can be confusing and against the "easier to use" vision we have.

For the default behavior, I now think that picking the first one only would be better as you suggested, to be consistent with what we return for `value()`. I'm open to more ideas / suggestions, though.

jaykorean commented on include/rocksdb/multi_cf_iterator.h:
```diff
@@ -0,0 +1,47 @@
+
+  virtual const AttributeGroups& attribute_groups() const = 0;
+};
+extern MultiCfIterator* NewMultiColumnFamilyIterator(
```
I've been following the same pattern as other iterator creations, but I just learned that this is not necessary. Thanks for letting me know.

jaykorean commented on include/rocksdb/wide_columns.h:
```diff
@@ -238,9 +238,17 @@ class AttributeGroup {
 
+inline bool operator==(const AttributeGroup& lhs, const AttributeGroup& rhs) {
+  return lhs.column_family() == rhs.column_family() &&
+         lhs.columns() == rhs.columns();
```
Columns are always sorted (by name) before added. Not sure if I understand your concern here. Could you provide a quick example?

pdillinger replied:
The data structure is fully defined in wide_columns.h. It is therefore reusable. This (reusable / general) comparison operator assumes an invariant on the data that the class itself does not describe. That mismatch (or omission) is asking for mismatch in expectations (potential bug).

Should this invariant always be true of WideColumns? I don't know, but that should be clear to users and developers.

pdillinger replied:
> > an invariant on the data that the class itself does not describe
>
> column names and values are always compared bytewise and it's commented in the class, but I think we can do better documentation.

Which class? The only reference I see to something like "compared bytewise" is on `operator==` for WideColumn, which is outside the class. There is no reference to sorting or ordering.

> Are you worried that it is not clear enough for users and developers to know that WideColumns are always sorted by column names

I don't see any documentation (APIs in db.h, wide_columns.h or wiki) indicating that sortedness is expected or provided, or whether perhaps the user ordering provided in PutEntity is preserved. I don't even see clear documentation of whether PutEntity completely overwrites a previous PutEntity on the same key or only on a column-by-column basis.

Nor exactly what Delete() does after PutEntity, just "works the same way as for regular key-values." So does it delete only the default column (since you can't specify a column)?

CC @ltamasi

> and both name and value of all columns should match for two WideColumns to be considered equal?

The presumption for `operator==` in C++ is structural equality. The question is what relationship that has to semantic equality, which can be different in various contexts.

ltamasi replied:
> I don't see any documentation (APIs in db.h, wide_columns.h or wiki) indicating that sortedness is expected or provided, or whether perhaps the user ordering provided in PutEntity is preserved. I don't even see clear documentation of whether PutEntity completely overwrites a previous PutEntity on the same key or only on a column-by-column basis.
> 
> Nor exactly what Delete() does after PutEntity, just "works the same way as for regular key-values." So does it delete only the default column (since you can't specify a column)?
> 
> CC @ltamasi

These points are mostly covered by the API comments at https://github.com/facebook/rocksdb/blob/main/include/rocksdb/db.h#L430-L459 , which, similarly to those of `Put`, use the concept of database entries associated with keys. After a `PutEntity`, the key is associated with that entity, regardless of any previous value (whether it's a plain KV, an existing entity, or nothing at all); after a `Delete`, the key is not associated with any database entry, again regardless of any previous value. There is no expectation that the input `WideColumns` structure is sorted, although it does get sorted by RocksDB, and thus anything read back from RocksDB will be sorted. We can extend the comments to clarify this point.

@jaykorean Sorry about the delay! Will review the PR shortly.

pdillinger replied:
@ltamasi If I was expecting PutEntity to do partial update, I might assume it's sloppy language, overly copying from Put(), even if it's technically correct and consistent. "will be overwritten" doesn't always mean entirely overwritten.

I think part of the confusion is singular "entry" and "entity" when dealing with plural things. Even if the description of Delete() is technically correct, it's not clear that it can deal with plurals like wide columns, especially since there's no DeleteEntity (like PutEntity / GetEntity). A reasonable user would doubt whether it handles new functionality even if its un-updated description technically implies it does.

jaykorean commented on include/rocksdb/db.h:
```diff
@@ -969,6 +971,16 @@ class DB {
+  // When the same key exists in more than one column families, the iterator
+  // selects the value from the first column family containing the key, in the
+  // order provided in the `column_families` parameter.
+  virtual MultiCfIterator* NewMultiCfIterator(
```
Do you mean we should return shared pointers in new APIs instead?

ltamasi replied:
In this case, `unique_ptr` would be the best choice I think. Shared pointers have both a memory and a CPU overhead (and clients can always create a `shared_ptr` later if needed).

ltamasi commented on db/multi_cf_iterator_impl.h:
```diff
@@ -0,0 +1,158 @@
+  }
+  const WideColumns& columns() const override {
+    assert(false);
+    // TODO - Lazily merge columns from child iterators
```
We could do something simpler here like returning the `columns()` of the first CF's iterator if the key is present in that CF, or an empty columns structure otherwise.

This would be similar to the invariant about `value()` returning the value of the default column from `columns()`, or an empty value if `columns()` doesn't include the default column. (BTW, regardless of how we choose to define `columns()`, we would want to make sure this invariant still holds.)

ltamasi replied:
> We could do something simpler here like returning the `columns()` of the first CF's iterator if the key is present in that CF, or an empty columns structure otherwise.

Another similar idea would be to expose the default column family's `columns()` here if the current key is present in the default CF, or an empty columns structure if it's not.

jaykorean commented on db/db_impl/db_impl.cc:
```diff
@@ -3992,6 +3996,28 @@ ArenaWrappedDBIter* DBImpl::NewIteratorImpl(
+  const Comparator* first_comparator = column_families[0]->GetComparator();
+
+  for (size_t i = 1; i < column_families.size(); ++i) {
+    if (first_comparator->GetId().compare(
```
Please let me know if there's a better way to do this.

pdillinger replied:
For efficiency, you probably want to try pointer == first, which should be a common case.

jaykorean commented:
Thank you all for the review.

I've addressed all the comments and made the following changes:

1. Removed all unnecessary `extern` functions.
2. Updated the API to return a unique pointer instead of a raw pointer.
3. Updated the API to no longer take the comparator as a parameter and instead use the first comparator from the column families. Iterator creation fails if the comparator does not match comparators in all other CFs.
4. Updated `columns()` to now return the `columns()` from the first CF that has the key (maintaining the same invariant as value()).
5. Added unit tests to verify points 3 and 4.

Please take another look when you get a chance.

jaykorean commented:
Thank you all for the review.

I've addressed all the comments and made the following changes:

1. Removed all unnecessary extern functions.
2. Updated the API to return a unique pointer instead of a raw pointer.
3. Updated the API to no longer take the comparator as a parameter and instead use the first comparator from the column families. Iterator creation fails if the comparator does not match comparators in all other CFs.
4. Updated columns() to now return the columns() from the first CF that has the key (maintaining the same invariant as value()).
5. Added unit tests to verify points 3 and 4.

Please take another look when you get a chance.

pdillinger commented on db/db_impl/db_impl.cc:
```diff
@@ -3720,6 +3721,28 @@ ArenaWrappedDBIter* DBImpl::NewIteratorImpl(
+std::unique_ptr<MultiCfIterator> DBImpl::NewMultiCfIterator(
+    const ReadOptions& _read_options,
+    const std::vector<ColumnFamilyHandle*>& column_families) {
+  assert(column_families.size() > 0);
```
In public APIs, at least if a precondition is easy to check, I believe we prefer to check it even in release builds and return an appropriate error (rather than crashing).

pdillinger replied:
(Complex preconditions like "these keys are sorted" would cost substantially in performance to check)

pdillinger reviewed (changes requested):
Looking good. Will probably stamp after this round. 👍

pdillinger commented on db/db_impl/db_impl.cc:
```diff
@@ -3720,6 +3721,31 @@ ArenaWrappedDBIter* DBImpl::NewIteratorImpl(
+    const std::vector<ColumnFamilyHandle*>& column_families) {
+  if (column_families.size() < 2) {
+    return std::make_unique<EmptyMultiCfIterator>(
+        Status::InvalidArgument("Less than 2 CFs were provided"));
```
Hmm, what to do with a single CF is a tricky question. I initially though that InvalidArgument was a good idea to promote efficient use of the best API, but then I remembered that checking status from a returned iterator is not always done properly. It would be easy for a user to mix up the result provided here with "success but no data found," if they only check valid() and not status(). We don't encourage sloppy ignoring of Statuses, but we should not ignore that it happens (hahahaha).

Another possibility to consider is to just make multi-CF awareness part of plain Iterator and then NewMultiCfIterator() with a single CF could be exactly equivalent to NewIterator(), with a special case in this code.

Do we foresee adding things to the abstract MultiCfIterator that would be difficult or nonsensical to implement for a single-CF iterator?

pdillinger reviewed (commented):
Awesome. I think my only outstanding question is whether it's best to roll multi-CF awareness into the general Iterator interface, so that the experience can be more unified and appropriate implementations chosen automatically regardless of number of CFs being iterated.

jaykorean commented on include/rocksdb/iterator.h:
```diff
@@ -102,6 +103,12 @@ class Iterator : public Cleanable {
   }
 
+  // UNDER CONSTRUCTION - DO NOT USE
+  virtual const AttributeGroups& attribute_groups() const {
```
I plan to implement this in DBIter in a separate PR.

pdillinger reviewed (approved):
Please update the PR title and description (and re-import or copy to phabricator), but otherwise LGTM. (Assumes for now we are moving toward the unified iterator experience.) Thanks!
